### DISCUSSION TITLE
How do I update a field on a model but invoke an external service before successfully deleting it?

### DISCUSSION THREAD

**Discussion by kennarddh** [prisma/prisma#25465] [Q&A]
### Question

I want to update an article to change the `thumbnailImageFileName`, this is the file name of an image stored in external service (minio). How do I to delete the previous image, upload the new image, and update the database while maintaining consistency. To prevent for example the previous image got deleted but the new image fails to be added or the image got replaced but the prisma fails to update or the article row in database has already been updated while this request update process is happening.

I'm thinking like this

```ts
try {
  const article = await prisma.articles.findFirst({ where: { id: request.params.id }})

  try {
    const newThumbnailImageFileName = await uploadImage(imageBuffer)

    await deleteImage(article.thumbnailImageFileName)

    try {
      await prisma.articles.update({ where: { id: request.params.id }, data: {} })
    } catch {
      // handle update error
    }
  } catch {
    // handle images error
  }
} catch {
  // handle select error
}
```

What if the row got updated by another request while this request is deleting/uploading the image.

How do I prevent such race condition and make sure it success or fail as a whole, without dangling image stored and not used. 

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
generator client {
    provider        = "prisma-client-js"
    binaryTargets   = ["native", "linux-musl-openssl-3.0.x"]
    previewFeatures = ["fullTextSearch", "fullTextIndex"]
}

datasource db {
    provider = "mysql"
    url      = env("DATABASE_URL")
}

model Articles {
    id Int @id @default(autoincrement())

    title   String @db.VarChar(256)
    content String @map("description") @db.Text

    createdAt DateTime @default(now())
    updatedAt DateTime @updatedAt

    thumbnailImageFileName String?          @db.VarChar(256)
}
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: Windows
- Database: MariaDB
- Node.js version: 20.12.2

- Run `prisma -v` to see your Prisma version and paste it
Environment variables loaded from .env
prisma                  : 5.18.0
@prisma/client          : 5.18.0
Computed binaryTarget   : windows
Operating System        : win32
Architecture            : x64
Node.js                 : v20.12.2
Query Engine (Node-API) : libquery-engine 4c784e32044a8a016d99474bd02a3b6123742169 (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Schema Engine           : schema-engine-cli 4c784e32044a8a016d99474bd02a3b6123742169 (at node_modules\@prisma\engines\schema-engine-windows.exe)
Schema Wasm             : @prisma/prisma-schema-wasm 5.18.0-25.4c784e32044a8a016d99474bd02a3b6123742169
Default Engines Hash    : 4c784e32044a8a016d99474bd02a3b6123742169
Studio                  : 0.502.0
Preview Features        : fullTextIndex, fullTextSearch

**Comment by ludralph** [comment]
Hi @kennarddh 👋 

You may want to take a look at [Optimistic Concurrency Control ](https://www.prisma.io/docs/orm/prisma-client/queries/transactions#optimistic-concurrency-control) for handling the update of an article's `thumbnailImageFileName` while ensuring consistency and preventing race conditions. 

You can use a `version` field in your database model to track changes and ensure that updates are only applied if another process has not modified the data since it was last read.
```prisma
model Articles {
    id Int @id @default(autoincrement())

    title   String @db.VarChar(256)
    content String @map("description") @db.Text

    createdAt DateTime @default(now())
    updatedAt DateTime @updatedAt

    thumbnailImageFileName String? @db.VarChar(256)
    version Int
}
```
Use the `version` field to ensure that the update is only applied if the data has not been modified by another process.

```js
async function updateArticleThumbnail(request, imageBuffer) {
  try {
    const article = await prisma.articles.findFirst({
      where: { id: request.params.id }
    });

    if (!article) {
      throw new Error('Article not found');
    }

    const newThumbnailImageFileName = await uploadImage(imageBuffer);

    try {
      await prisma.articles.update({
        where: { id: request.params.id, version: article.version },
        data: {
          thumbnailImageFileName: newThumbnailImageFileName,
          version: { increment: 1 }
        }
      });

      await deleteImage(article.thumbnailImageFileName);
    } catch (error) {
      // Handle update error, possibly revert the image upload
      await deleteImage(newThumbnailImageFileName);
      throw error;
    }
  } catch (error) {
    // Handle select error or image upload error
    console.error(error);
  }
}
```
The update operation is only executed if the version field matches the version of the article when it was first read. If another process has updated the article in the meantime, the update will fail, and you can handle this case appropriately, such as by retrying the operation or notifying the user.

**Reply by kennarddh** [reply]
Thanks!

**Reply by kennarddh** [reply]
@ludralph When `prisma.articles.update` throws error because of the version is mismatched, how do I differentiate the error with when the article is already deleted or not found. Both throws P2016 without any additional metadata.

**Reply by ludralph** [reply]
You're correct that both a version mismatch and a record not found scenario can result in a P2016 error. Prisma does not provide a way to differentiate between these cases directly from the error code alone. We can implement a workaround to distinguish between these scenarios. Here's an approach you can use:

1. First, attempt to find the article.
2. If the article is found, proceed with the update.
3. If the update fails, it's likely due to a version mismatch.

```js
async function updateArticleThumbnail(articleId: string, imageBuffer: Buffer) {
  return await prisma.$transaction(async (tx) => {
    // 1. Find the article
    const article = await tx.article.findUnique({
      where: { id: articleId },
      select: { id: true, thumbnailImageFileName: true, version: true }
    })

    if (!article) {
      throw new Error('Article not found')
    }

    // 2. Upload the new image
    const newThumbnailImageFileName = await uploadImage(imageBuffer)

    try {
      // 3. Attempt to update the article
      const updatedArticle = await tx.article.update({
        where: { 
          id: articleId,
          version: article.version
        },
        data: {
          thumbnailImageFileName: newThumbnailImageFileName,
          version: { increment: 1 }
        }
      })

      // 4. If update succeeds, delete the old image
      await deleteImage(article.thumbnailImageFileName)

      return updatedArticle
    } catch (error) {
      // Delete the newly uploaded image
      await deleteImage(newThumbnailImageFileName)

      if (error instanceof Prisma.PrismaClientKnownRequestError) {
        if (error.code === 'P2016') {
          // Since we've already checked for existence, this is likely a version mismatch
          throw new Error('Concurrent update detected. Please try again.')
        }
      }
      throw error
    }
  })
}
```
We first check if the article exists using `findUnique`. If it doesn't, we throw a "Article not found" error. If the article is found, we proceed with the update attempt. If the update fails with a P2016 error, we can be reasonably certain that it's due to a version mismatch, since we've already confirmed the article's existence. This approach allows you to differentiate between the "not found" and "version mismatch" scenarios, even though they both result in the same error code.

Remember to handle these different error types appropriately in the calling code:
```js
try {
  const updatedArticle = await updateArticleThumbnail(articleId, imageBuffer)
  // Handle successful update
} catch (error) {
  if (error.message === 'Article not found') {
    // Handle not found case
  } else if (error.message === 'Concurrent update detected. Please try again.') {
    // Handle version mismatch case
  } else {
    // Handle other errors
  }
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
